### ISSUE TITLE
SpelEvaluationException is getting while evaluating partitionKeyExpression config

### ISSUE THREAD

**Issue by msshitole** [spring-cloud/spring-cloud-stream#1877]
In the latest release ( Horsham.SR1), the default config for evaluating` partitionKeyExpression` is not working. Its throwing `SpelEvaluationException` which was working fine in the previous version.
**`spring.cloud.stream.bindings.emit-out-0.producer.partitionKeyExpression=payload._id`**

Here is the exception stacktrace:
```
[ERROR] 2019-12-20T12:58:57,741 - [app-source] org.springframework.integration.handler.LoggingHandler - org.springframework.messaging.MessageDeliveryException: failed to send Message to channel 'emit-out-0'; nested exception is org.springframework.expression.spel.SpelEvaluationException: EL1008E: Property or field '_id' cannot be found on object of type 'byte[]' - maybe not public or not valid?, failedMessage=GenericMessage [payload=byte[1807], headers={contentType=avro/bytes, id=b25a26c6-3479-abcdf-eaf4-11da5c3f5fd4, timestamp=1576843137737}]
	at org.springframework.integration.support.utils.IntegrationUtils.wrapInDeliveryExceptionIfNecessary(IntegrationUtils.java:166)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:483)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:403)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:187)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:166)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:109)
	at org.springframework.integration.router.AbstractMessageRouter.doSend(AbstractMessageRouter.java:206)
	at org.springframework.integration.router.AbstractMessageRouter.handleMessageInternal(AbstractMessageRouter.java:188)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:170)
	at org.springframework.integration.handler.AbstractMessageHandler.onNext(AbstractMessageHandler.java:219)
	at org.springframework.integration.handler.AbstractMessageHandler.onNext(AbstractMessageHandler.java:57)
	at org.springframework.integration.endpoint.ReactiveStreamsConsumer$DelegatingSubscriber.hookOnNext(ReactiveStreamsConsumer.java:165)
	at org.springframework.integration.endpoint.ReactiveStreamsConsumer$DelegatingSubscriber.hookOnNext(ReactiveStreamsConsumer.java:148)
	at reactor.core.publisher.BaseSubscriber.onNext(BaseSubscriber.java:160)
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:123)
	at reactor.core.publisher.EmitterProcessor.drain(EmitterProcessor.java:426)
	at reactor.core.publisher.EmitterProcessor.onNext(EmitterProcessor.java:268)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.drain(FluxCreate.java:793)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.next(FluxCreate.java:718)
	at reactor.core.publisher.FluxCreate$SerializedSink.next(FluxCreate.java:153)
	at org.springframework.integration.channel.FluxMessageChannel.doSend(FluxMessageChannel.java:63)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:453)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:403)
	at org.springframework.integration.channel.FluxMessageChannel.lambda$subscribeTo$2(FluxMessageChannel.java:83)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:189)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runAsync(FluxPublishOn.java:398)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:484)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.springframework.expression.spel.SpelEvaluationException: EL1008E: Property or field '_id' cannot be found on object of type 'byte[]' - maybe not public or not valid?
	at org.springframework.expression.spel.ast.PropertyOrFieldReference.readProperty(PropertyOrFieldReference.java:217)
	at org.springframework.expression.spel.ast.PropertyOrFieldReference.getValueInternal(PropertyOrFieldReference.java:104)
	at org.springframework.expression.spel.ast.PropertyOrFieldReference.access$000(PropertyOrFieldReference.java:51)
	at org.springframework.expression.spel.ast.PropertyOrFieldReference$AccessorLValue.getValue(PropertyOrFieldReference.java:406)
	at org.springframework.expression.spel.ast.CompoundExpression.getValueInternal(CompoundExpression.java:92)
	at org.springframework.expression.spel.ast.SpelNodeImpl.getValue(SpelNodeImpl.java:112)
	at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:330)
	at org.springframework.cloud.stream.binder.PartitionHandler.extractKey(PartitionHandler.java:109)
	at org.springframework.cloud.stream.binder.PartitionHandler.determinePartition(PartitionHandler.java:90)
	at org.springframework.cloud.stream.binding.MessageConverterConfigurer$PartitioningInterceptor.preSend(MessageConverterConfigurer.java:402)
	at org.springframework.integration.channel.AbstractMessageChannel$ChannelInterceptorList.preSend(AbstractMessageChannel.java:613)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:443)
```

Just FYI, the workaround is, implementing `PartitionKeyExtractorStrategy` and with the following config. 
**spring.cloud.stream.bindings.emit-out-0.producer.partitionKeyExtractorName=customPartitionKeyExtractor**
Could you please check this once. Thank you!

**Comment by olegz** [comment]
Actually I don't think this is an issue. You can not have  expressions on the payload unless your expression evaluates a `byte[]` since payload is ***always*** `byte[]`. We actually talk about it extensively in [content-type-management](https://cloud.spring.io/spring-cloud-static/spring-cloud-stream/3.0.1.RELEASE/reference/html/spring-cloud-stream.html#content-type-management) section where we talk about _wire format_. 

The bottom line is the payload is converted right before the actual handler is invoked based on the input type of the handler. Until then it's byte[]. The fact that it may have worked before  is actually a bug IMHO.

**Comment by olegz** [comment]
Wait, I take that back. YOu're talking about _output_. That's different, let me look

**Comment by olegz** [comment]
Ok, so I think I am back to my original statement. But let me explain. . .

Just as I stated earlier, inputs are not converted up until the last moment (when input type is known). And with functions it is even more relevant now since type conversion is no longer in spring-cloud-stream, but in spring-cloud-function. The same goes for output where payload get's converted back to the _wire format_ as soon as function is invoked and output is produced. So, that is why you see this exception

The reason why it may have worked before is because there was a [bug](https://github.com/spring-cloud/spring-cloud-stream/issues/1855) where we completely excluded MessageConverterConfigurer for functions which is now fixed.

So while it will still work with annotation-based listeners, I strongly believe that `payload` should be off limits for SpEL both on the input and output. We have headers for this. Think about it this way. the mailman makes their decision about delivery of the letter not by reading your letter (payload) but rather only the information available on the envelope (headers).

So, I am going to keep this issue open as I assume more people will notice that. But what we're going to fix is update documentation and possible blog-post explaining what I just explained to you.

That said, please feel free to express your opinion, regardless. . .

**Comment by msshitole** [comment]
OK. Fair enough. 

Please correct my understanding. To enable **partitioning** in output binding, below configurations are required.
**spring.cloud.stream.bindings.emit-out-0.producer.partitionKeyExtractorName=customPartitionKeyExtractor
spring.cloud.stream.bindings.emit-out-0.producer.partitionCount=6**

And here **key** is extracted from the payload and not from the header. Here is the implementation of bean `customPartitionKeyExtractor` :
```
public class CustomPartitionKeyExtractor implements PartitionKeyExtractorStrategy {

    @Autowired
    private MessageConverter avroDocumentMessageConverter;

    @Override
    public String extractKey(Message<?> message) {
        Document document = (Document) avroDocumentMessageConverter.fromMessage(message, Document.class);
        return document.getId$1();
    }
}
```

Now the question is, the above configuration and its respective bean - will it enable the partitioning in an expected way or not? is it applied? and it's a legitimate way?

is this OK for now? or need another way (header-based)... will it be fine? or something else to be needed on the same. So then accordingly needs an update to the documentation.

**Comment by olegz** [comment]
Not necessarily. You can communicate this meta-information through message headers. For example, if you want to communicate some meta-information from your function then have the output type of the function as Message<POJO>.

```
public Function<Blah, Message<POJO>> func() {
       return v -> {
               . . . . .
               return MessageBuilder.withPayload(..).setHeader("payloadId", id).build();
       };
```

Then you can simply do `spring.cloud.stream.bindings.emit-out-0.producer.partitionKeyExpression=headers.payloadId`

It is the same on the input where if you are interested in certain meta information then define your input type as Message and we'll pass it to you.

**Comment by msshitole** [comment]
OK. But then both these ways are legitimate and will work. 
I hope my understanding is correct.

In case if you are thinking about deprecating the **customPartitionKeyExtractor**. Then we have to go on the setting headers only.

or currently, in (Horhsam.SR1), partitioning will only work on the **partitionKeyExpression** config?

**Comment by olegz** [comment]
No, there are no plans on deprecating `customPartitionKeyExtractor` at all.
Yes both ways are legitimate.

**Comment by msshitole** [comment]
A sigh of relief! The less todo is more better for us. Then we can only concentrate on the business logic and **spring** will take care of the rest. ;)

**Comment by olegz** [comment]
Duplicate to https://github.com/spring-cloud/spring-cloud-stream/issues/1893
